client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -162,7 +162,14 @@ private String updateGroupStateIfNeeded() throws ReinitializationRequiredExcepti
             atCheckpoint = groupState.getCheckpoint();
             if (atCheckpoint != null) {
                 log.info("{} at checkpoint {}", this, atCheckpoint);
-                return atCheckpoint;
+                if (groupState.isCheckpointSilent(atCheckpoint)) {
+                    // Checkpoint the reader immediately with the current position. Checkpoint Event is not generated.
+                    groupState.checkpoint(atCheckpoint, getPosition());
```

Comment:
The implementation of this method updates the Positions in assignedSegments the reader group state. I don't this is a desired side effect.

Reply:
Updating the offsets in state.assignedSegments will cause a different offset to be used in the event a reader goes offline. The current contract is that this is based on the last checkpoint. (or the position provided). As this is not an exposed checkpoint this violates that contract.

Reply:
A couple of questions here:

1- I'd like to clarify the contract that we need to satisfy. The contract that @tkaitchuck is referring to is with respect to visible only (triggered by the application or automatic) or all checkpoints? This PR is adding silent checkpoints, so I wonder whether the contract should take such checkpoints into account. 
2- In the case that we update offsets as @shrids proposes, is the concern that we don't know whether the reader has processed any event since the last checkpoint, and as such, we conservatively reset to the position of the last checkpoint? I also assume that if we provide a `Position` object to the reader offline call, then it doesn't need to match the offsets of a checkpoint.

Reply:
1- If an application invokes `readerOffline` and passes null for the position, the contract we have is that it resumes from the previous checkpoint. I don't think this should count, because it is not a checkpoint.
2- Correct.

Reply:
done.